### PR TITLE
[Merged by Bors] - feat: quadratic forms applied to sums decompose into quadratic and bilinear parts

### PR DESCRIPTION
**Pull request by eric-wieser**
---

- [x] depends on: #14212 

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by eric-wieser** [comment]
The eventual motivation here is something like
```lean
noncomputable def QuadraticForm.toTriMatrix {ι R M} [CommRing R] [AddCommGroup M] [Module R M]
    (Q : QuadraticForm R M) (bm : Basis ι R M) [LinearOrder ι] : Matrix ι ι R :=
  Matrix.of fun i j =>
    ltByCases i j
        (fun _ => polar Q (bm i) (bm j))
        (fun _ => Q (bm i))
        (fun _ => 0)


theorem QuadraticForm.toTriMatrix_mul  {ι R M} [CommRing R] [AddCommGroup M] [Module R M]
    (Q : QuadraticForm R M) (bm : Basis ι R M) [LinearOrder ι] [Fintype ι] :
    (Matrix.toBilin bm (Q.toTriMatrix bm)).toQuadraticForm = Q := sorry
```
where the `ltByCases` is materializing the sum over `sym2`.

File: Mathlib/LinearAlgebra/QuadraticForm/Basic.lean [hunk 1]
```diff
@@ -360,6 +361,39 @@ theorem choose_exists_companion : Q.exists_companion.choose = polarBilin Q :=
       add_sub_cancel_left]
 #align quadratic_form.some_exists_companion QuadraticForm.choose_exists_companion
 
+protected theorem map_add (Q : QuadraticForm R M) (x y : M):
+    Q (x + y) = Q x + Q y + polar Q x y := by
+  simp [polar]
+
+protected theorem map_sum {ι} [DecidableEq ι] (Q : QuadraticForm R M) (s : Finset ι) (f : ι → M) :
+    Q (∑ i ∈ s, f i) = ∑ i ∈ s, Q (f i) +
+      ∑ ij in s.sym2.filter (¬ ·.IsDiag),
```
[the remark this answers is not in the document]
Comment (YaelDillies) [reply Mathlib/LinearAlgebra/QuadraticForm/Basic.lean:370]: A bit of a hack, but you can write that as 
```suggestion
      ∑ ij in (s.powerset.slice 2).attach.map
        ⟨fun t ↦ (Sym2.equivSym _).symm ⟨t.1.1, (mem_slice.1 t.2).2⟩,
          (Sym2.equivSym _).symm.injective.comp sorry /- good luck -/⟩,
```

[the remark this answers is not in the document]
Comment (eric-wieser) [reply Mathlib/LinearAlgebra/QuadraticForm/Basic.lean:370]: I think I'd prefer to keep the `DecidableEq` in that case!

[the remark this answers is not in the document]
Comment (urkud) [reply Mathlib/LinearAlgebra/QuadraticForm/Basic.lean:384]: If you want to be computationally efficient, then you need to implement `List.subDiagonal` and use it to define `Finset.sym2OffDiagonal`. For now, I suggest that we use `DecidableEq`.

[the remark this answers is not in the document]
Comment (eric-wieser) [reply Mathlib/LinearAlgebra/QuadraticForm/Basic.lean:384]: List.subDiagonal exists as `List.diagM` I think.

File: Mathlib/LinearAlgebra/QuadraticForm/Basic.lean [hunk 2]
```diff
@@ -379,6 +380,35 @@ theorem choose_exists_companion : Q.exists_companion.choose = polarBilin Q :=
       add_sub_cancel_left]
 #align quadratic_form.some_exists_companion QuadraticMap.choose_exists_companion
 
+protected theorem map_sum {ι} [DecidableEq ι] (Q : QuadraticMap R M N) (s : Finset ι) (f : ι → M) :
+    Q (∑ i ∈ s, f i) = ∑ i ∈ s, Q (f i) +
+      ∑ ij in s.sym2.filter (¬ ·.IsDiag),
+        Sym2.lift ⟨fun i j => polar Q (f i) (f j), fun _ _ => polar_comm _ _ _⟩ ij := by
```
Comment (YaelDillies) [inline Mathlib/LinearAlgebra/QuadraticForm/Basic.lean:386]: nit
```suggestion
        Sym2.lift ⟨fun i j => polar Q (f i) (f j), fun _ _ => polar_comm ..⟩ ij := by
```

**Review by YaelDillies — COMMENTED** [review]
Looks otherwise good to me

File: Mathlib/LinearAlgebra/QuadraticForm/Basic.lean (hunk 2 above)
Comment (eric-wieser) [reply Mathlib/LinearAlgebra/QuadraticForm/Basic.lean:386]: I don't really find this an improvement, it makes more work for the reader and the elaborator

**Comment by YaelDillies** [comment]
maintainer merge

**Comment by mattrobball** [comment]
!bench

**Comment by mattrobball** [comment]
bors merge

**Pull request closed without merging**